### PR TITLE
performance_schema: revised support for SELECT

### PR DESCRIPTION
**Pull request by nieyy**
hard to split into 2 PRs for `SELECT` only, revised support for `SELECT` functions:
1. using goleveldb instead of map only for concurrent considerations (INSERT/DELETE/UPDATE/TRUNCATE).
2. code refactoring.

### CODE REVIEW

File: perfschema/init.go [hunk 1]
```diff
@@ -0,0 +1,1411 @@
...
+// 		ROLE			CHAR(16) NOT NULL  DEFAULT '%',
+// 		ENABLED			ENUM('YES','NO') NOT NULL  DEFAULT 'YES',
+// 		HISTORY			ENUM('YES','NO') NOT NULL  DEFAULT 'YES');
+func (ps *perfSchema) buildSetupActorsModel() {
+	tbName := TableSetupActors
+	table := []struct {
+		tp    byte
+		size  int
+		flag  uint
+		def   interface{}
```
Comment (qiuyesuifeng) [inline perfschema/init.go]: It is better to use default instead, def seems like definition.

Comment (nieyy) [inline perfschema/init.go]: no problem

File: perfschema/init.go [hunk 2]
```diff
@@ -0,0 +1,1411 @@
...
+)
+
+// CREATE TABLE if not exists performance_schema.setup_actors (
+// 		HOST			CHAR(60) NOT NULL  DEFAULT '%',
+// 		USER			CHAR(32) NOT NULL  DEFAULT '%',
+// 		ROLE			CHAR(16) NOT NULL  DEFAULT '%',
+// 		ENABLED			ENUM('YES','NO') NOT NULL  DEFAULT 'YES',
+// 		HISTORY			ENUM('YES','NO') NOT NULL  DEFAULT 'YES');
+func (ps *perfSchema) buildSetupActorsModel() {
+	tbName := TableSetupActors
```
Comment (qiuyesuifeng) [inline perfschema/init.go]: How about extract to a function?
Seems u can use table, cols as variables, then the code will be shorter and clearer.

Comment (nieyy) [inline perfschema/init.go]: I can do that, any other comments?

File: perfschema/init.go [hunk 3]
```diff
@@ -0,0 +1,1411 @@
...
+	table := []struct {
```
Comment (ngaut) [inline perfschema/init.go]: We can defined it as named structure. DRY!

Comment (nieyy) [inline perfschema/init.go]: Good suggestion, thanks

**Comment by ngaut** [comment]
PTAL @shenli @coocood @qiuyesuifeng

File: perfschema/perfschema.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+// PerfHandle is the only access point for the in-memory performance schema information
+var PerfHandle PerfSchema
+
+func init() {
+	// TODO: need error handling later
+	schema := &perfSchema{}
+	_ = schema.initialize()
+
+	var v interface{} = schema
```
Comment (shenli) [inline perfschema/perfschema.go]: Is "PerfHandle = schema" ok? Why introduce the v?

File: perfschema/perfschema.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+type PerfSchema interface {
+	// For SELECT statement only.
+	NewPerfSchemaPlan(tableName string) (plan.Plan, error)
+}
+
+type perfSchema struct {
+	tables map[string]*model.TableInfo
+	fields map[string][]*field.ResultField
+	stores map[string]*leveldb.DB
+	lsns   map[string]*uint64
```
Comment (shenli) [inline perfschema/perfschema.go]: What is this 'lsns' used for? Any comments?

File: perfschema/perfschema.go (hunk 4 above)
Comment (nieyy) [inline perfschema/perfschema.go]: compilation error, I tried it and Google, and fix it, :)

File: perfschema/perfschema.go (hunk 5 above)
Comment (nieyy) [inline perfschema/perfschema.go]: lsn is a DB word, LSN = Logical Sequence Number, now it's just a simplified version, each time we insert a new record, we just increment LSN.

**Comment by shenli** [comment]
LGTM

File: perfschema/init.go [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+		{mysql.Enum{Name: "EVENT", Value: 1}, "%", "%", mysql.Enum{Name: "YES", Value: 1}, mysql.Enum{Name: "YES", Value: 1}},
+		{mysql.Enum{Name: "FUNCTION", Value: 2}, "mysql", "%", mysql.Enum{Name: "NO", Value: 2}, mysql.Enum{Name: "NO", Value: 2}},
+		{mysql.Enum{Name: "FUNCTION", Value: 2}, "performance_schema", "%", mysql.Enum{Name: "NO", Value: 2}, mysql.Enum{Name: "NO", Value: 2}},
+		{mysql.Enum{Name: "FUNCTION", Value: 2}, "information_schema", "%", mysql.Enum{Name: "NO", Value: 2}, mysql.Enum{Name: "NO", Value: 2}},
+		{mysql.Enum{Name: "FUNCTION", Value: 2}, "%", "%", mysql.Enum{Name: "YES", Value: 1}, mysql.Enum{Name: "YES", Value: 1}},
+		{mysql.Enum{Name: "TABLE", Value: 3}, "mysql", "%", mysql.Enum{Name: "NO", Value: 2}, mysql.Enum{Name: "NO", Value: 2}},
+		{mysql.Enum{Name: "TABLE", Value: 3}, "performance_schema", "%", mysql.Enum{Name: "NO", Value: 2}, mysql.Enum{Name: "NO", Value: 2}},
+		{mysql.Enum{Name: "TABLE", Value: 3}, "information_schema", "%", mysql.Enum{Name: "NO", Value: 2}, mysql.Enum{Name: "NO", Value: 2}},
+		{mysql.Enum{Name: "TABLE", Value: 3}, "%", "%", mysql.Enum{Name: "YES", Value: 1}, mysql.Enum{Name: "YES", Value: 1}},
```
Comment (ngaut) [inline perfschema/init.go]: s/"%"/`%`

File: perfschema/perfschema.go (hunk 4 above)
Comment (qiuyesuifeng) [inline perfschema/perfschema.go]: Can u directly using `PerfHandle, _ = schema.(PerfSchema)`?

Comment (ngaut) [inline perfschema/perfschema.go]: Can be assigned directly. I made an example: http://play.golang.org/p/mc3-_JJV8Y

Comment (nieyy) [inline perfschema/perfschema.go]: it works, thanks, this is called explicitly conversion, right?

File: perfschema/const.go [hunk 7]
```diff
@@ -58,21 +58,91 @@ var PerfSchemaTables = []string{
 }
 
 // ColumnSetupActors contains the column name definitions for table setup_actors, same as MySQL.
```
Comment (ngaut) [inline perfschema/const.go]: Do we need to export all of these variables ?

Comment (nieyy) [inline perfschema/const.go]: Table names, column names, maybe we will need it later when implementing instrumentation points, :)

**Comment by ngaut** [comment]
LGTM

**Comment by ngaut** [comment]
Thanks @nieyy

**Comment by nieyy** [comment]
Really appreciate ur careful review and quick response.

**Pull request merged**